src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs

```diff
@@ -41,6 +41,8 @@ public static class TextLoaderSaverCatalog
         /// A column may also have dense values followed by sparse values represented in this fashion. For example,
         /// a row containing "1 2 5 2:6 4:3" represents two dense columns with values 1 and 2, followed by 5 sparsely represented
         /// columns with values 0, 0, 6, 0, and 3. The indices of the sparse columns start from 0, even though 0 represents the third column.</param>
+        /// <param name="readMultilines">If true, it will accept new line characters inside quoted fields. Thus, enabling that a single field contain
+        /// multiple lines of text. If allowQuoting is false, this parameter is ignored.</param>
```

Comment:
Are there scenarios where we would want this to be false? 
If allowQuoting is true, don't we always want readMultiLines = true?

Reply:
You're right in that we wouldn't need an API change in that case. I'm hesitant to always set readMultiLines as true if allowQuoting is enabled because of the reasons I've pointed out here:
https://github.com/dotnet/machinelearning/pull/5125#discussion_r425351571

But yeah, in general it would be more convenient to simply use allowQuoting to decide if we want to read multilines, so that we can skip the part of having to create the readMultilines parameter.

So I'm not sure if we should do it or not. Any thoughts, @justinormont @harishsk ?

Reply:
How would we expect ModelBuilder to decide the values for allowQuoting and readMultilines? If it always sets them to true, then we are going to run into the bad case you mention in the discussion where the reader can potentially read the whole file. If they always set them to false, they won't be able to read a lot of files correctly. Do we expect them to try one different permutations till it succeeds?

---
In reply to: [425403877](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425403877) [](ancestors = 425403877)

Reply:
Yeah, it's certainly a dilemma, an I wouldn't know what the ModelBuilder team would prefer, so perhaps @LittleLittleCloud has something to say about this?

My recommendation is that ModelBuilder always uses allowQuoting and readMultilines. readMultilines should always work if the input file is correctly formatted, so in case an exception occurs in the case I pointed to, then it would be a problem with the dataset, and not with the tool. Perhaps adding an option for readMultiline to ModelBuilder's GUI could also work, but I don't know if it's possible to do it or convenient for the end user.

---
In reply to: [425464413](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425464413) [](ancestors = 425464413,425403877)

Reply:
@LittleLittleCloud @JakeRadMSFT, @justinormont We would greatly appreciate your input on this one.


---
In reply to: [425473921](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425473921) [](ancestors = 425473921,425464413,425403877)

Reply:
> harishsk 13 hours ago  Member
> How would we expect ModelBuilder to decide the values for allowQuoting and readMultilines? If it always sets them to true, then we are going to run into the bad case you mention in the discussion where the reader can potentially read the whole file. If they always set them to false, they won't be able to read a lot of files correctly. Do we expect them to try one different permutations till it succeeds?

The `TextLoader` option are swept over in the AutoML code, and the option-set which produces the most consistent result is used (w/ min threshold of 98%). Currently { allowSparse, allowQuote, separatorCandidates } is swept over in priority order to allow for tie breaking towards the more common/safe option. We would want to add `readMultilines` as another sweep option.

https://github.com/dotnet/machinelearning/blob/e50c4d20012e0d62852f404ae443afca7dad043e/src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs#L51-L78))

Reply:
As I've pointed out in [this comment](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425726377), the code you're pointing to seems to only  affect the TrySplitColumns() method. So, for instance, it seems that more changes will also be needed in all the methods of [ColumnInferenceApi](https://github.com/dotnet/machinelearning/blob/e50c4d20012e0d62852f404ae443afca7dad043e/src/Microsoft.ML.AutoML/ColumnInference/ColumnInferenceApi.cs) and whichever other methods that call them. Furthermore, perhaps SplitColumns() and ColumnInference() aren't the only places were changes are required to also use the new readMultilines option.

So I've talked with @harishsk about this, and we've decided that I won't address any changes to the AutoML API, and will focus now on solving the other issues and feature requests that @justinormont has asked on this PR. After this PR gets merged further work will be needed to surface this new feature on the AutoML API so it can be accessed by ModelBuilder. Because of this, I will revert my changes on the AutoML code that hardcoded `readMultiline = true` and remove also the test I had for ColumnInference /cc:@LittleLittleCloud

---
In reply to: [425743577](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425743577) [](ancestors = 425743577)

Reply:
Also, I wonder, is sweeping the readMultilines option good with ModelBuilder? I'm not sure how all of these work in AutoML/ModelBuilder, but if we actually need to train and evaluate a model for every sweeping option there is, then sweeping on readMultilines will double the time required to do get results, right? Since ModelBuilder works by letting the user set a time limit to the training process, would doubling the time be considered an issue?

---
In reply to: [425743577](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425743577) [](ancestors = 425743577)

Reply:
It's very quick to try and open the dataset w/ various parameters for the `TextLoader`. It tries to read the 1st 1000 rows of data; no model training/eval.

I believe we currently try up to 16 combination, adding this flag will move it to 32. The 16 is currently trivially fast, and should remain so. 

One corner case to perf test is the run-away quoted option, where the close-quote is never provided. I expect this won't impact actual users the run-away is stopped by the next double-quote, and the further lines will be misread causing that option set to not be successful (which is good).

We test the given `TextLoader` options set in `TryParseFile()`: _(no training/eval)_
https://github.com/dotnet/machinelearning/blob/e50c4d20012e0d62852f404ae443afca7dad043e/src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs#L81-L122

Reply:
> It's very quick to try and open the dataset w/ various parameters for the `TextLoader`. It tries to read the 1st 1000 rows of data; no model training/eval.
> 
> I believe we currently try up to 16 combination, adding this flag will move it to 32. The 16 is currently trivially fast, and should remain so.
> 
> One corner case to perf test is the run-away quoted option, where the close-quote is never provided. I expect this won't impact actual users the run-away is stopped by the next double-quote, and the further lines will be misread causing that option set to not be successful (which is good).
> 
> We test the given `TextLoader` options set in `TryParseFile()`: _(no training/eval)_
> https://github.com/dotnet/machinelearning/blob/e50c4d20012e0d62852f404ae443afca7dad043e/src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs#L81-L122

I agree with Justin, just adding another sweep option should be enough to solve parsing error.

Besides, I would not make any changes to AutoML.Net `ColumnInference` API, to me `readMultiline` is a trivial and  confusing detail that should be concealed from AutoML users

Reply:
Ok, so after discussing it offline with @LittleLittleCloud , we decided that model builder team will take care of changing anything necessary in AutoML.NET to make this work for them , once this PR is merged. 😄 

About the ColumnInference API, I thought it was important for you guys since the original issue of all of these ( #4460 ) explicitly talks about having problems with using ColumnInference() with files with new line characters inside quoted fields. But if it's not important, then only changing TryParseFile() might be enough and straightforward. In any case, I'll leave this discussion and fix up to the model builder team. Thanks.

Reply:
BTW, Just out of curiosity, does ModelBuilder uses the ColumnInference() method @LittleLittleCloud ?

Reply:
> BTW, Just out of curiosity, does ModelBuilder uses the ColumnInference() method @LittleLittleCloud ?
Yes, we use it in `AutoMLService`